python/tank/bootstrap/manager.py

```diff
@@ -236,32 +244,153 @@ def bootstrap_engine(self, engine_name, entity=None):
         the engine may not be the same as the API version that was
         executed during the bootstrap.
 
-        :param entity: Shotgun entity to launch engine for
-        :type entity: Dictionary with keys type and id
-        :param engine_name: name of engine to launch (e.g. ``tk-nuke``)
-        :returns: :class:`sgtk.platform.Engine` instance
+        :param engine_name: Name of engine to launch (e.g. ``tk-nuke``).
+        :param entity: Shotgun entity to launch engine for.
+        :returns: :class:`sgtk.platform.Engine` instance.
         """
-        log.info("Bootstrapping into engine %s for entity %s." % (engine_name, entity))
 
-        log.debug("Bootstrapping into environment.")
+        log.info("Bootstrapping engine %s for entity %s." % (engine_name, entity))
+
         tk = self._bootstrap_sgtk(engine_name, entity)
 
-        log.debug("Resolving context.")
-        if entity is None:
-            ctx = tk.context_empty()
+        engine = self._start_engine(tk, engine_name, entity)
+
+        return engine
+
+    def bootstrap_engine_async(self,
+                               engine_name,
+                               entity=None,
+                               progress_callback=None,
+                               completed_callback=None,
+                               failed_callback=None):
+        """
+        Create an :class:`sgtk.Sgtk` instance for the given engine and entity,
+        then launch into the given engine.
+
+        The :class:`sgtk.Sgtk` instance will be bootstrapped asynchronously in a background thread,
+        followed by launching the engine synchronously in the main application thread.
+        This will allow the main application to continue its execution and
+        remain responsive when bootstrapping the toolkit involves downloading files and
+        installing apps from the toolkit app store.
+
+        If entity is None, the method will bootstrap into the site
+        config. This method will attempt to resolve the config according
+        to business logic set in the associated resolver class and based
+        on this launch a configuration. This may involve downloading new
+        apps from the toolkit app store and installing files on disk.
+
+        3 callback functions can be provided.
+
+        A callback function that reports back on the bootstrap progress
+        with the following signature::
+
+            progress_callback(progress_value, message, current_index, maximum_index)
```

Comment:
From a API user point of view, it seems confusing that there are both a `progress_callback` param here and a `Manager.set_progress_callback()`.

It feels a little odd to me that we have two mechanisms to do the same thing! Why can’t we just update the function callback signature for `Manager.set_progress_callback()` to take the new `progress_value` parameter, and use your backwards compatibility method signature magic there instead? It seems the progress reporting should work the same way regardless of whether you are sync or async! 

I think the progress callback makes sense both in the sync and async case, so it kind of makes sense that it is a property on the manager class like it is today. But the on_complete/on_error callbacks are really only relevant when you run async.

So my suggestion would probably be to just have this method signature:

```
def bootstrap_engine_async(self,
                           engine_name,
                           entity=None,
                           completed_callback=None,
                           failed_callback=None):
```

calling code would then call

```python
# sync case
mgr.set_progress_callback(progress_cb)
try:
   e = mgr. bootstrap_engine("tk-maya")
except:
   print "Something went wrong"

# async case
mgr.set_progress_callback(progress_cb)
mgr.bootstrap_engine_async("tk-maya", complete_cb, error_cb)
```